File: src/chapkit/data/geo.py
"""Geospatial helpers for chapkit services."""

from collections.abc import Iterator
from typing import Any

# A GeoJSON bounding box in spec order: [west, south, east, north].
BoundingBox = tuple[float, float, float, float]


def _iter_positions(coordinates: Any) -> Iterator[tuple[float, float]]:
    """Yield every (lon, lat) position from an arbitrarily nested coordinate array."""
    if (
        isinstance(coordinates, (list, tuple))
        and len(coordinates) >= 2
        and isinstance(coordinates[0], (int, float))
        and isinstance(coordinates[1], (int, float))
    ):
        yield (float(coordinates[0]), float(coordinates[1]))
        return
    if isinstance(coordinates, (list, tuple)):
        for item in coordinates:
            yield from _iter_positions(item)


def _geometry_positions(geometry: Any) -> Iterator[tuple[float, float]]:
    """Yield every position from a GeoJSON geometry, recursing into collections."""
    if not isinstance(geometry, dict):
        return
    if geometry.get("type") == "GeometryCollection":
        for sub in geometry.get("geometries") or []:
            yield from _geometry_positions(sub)
        return
    yield from _iter_positions(geometry.get("coordinates"))


def _as_dict(feature_collection: Any) -> dict[str, Any]:
    """Coerce a geojson-pydantic model or mapping into a plain dict."""
    if hasattr(feature_collection, "model_dump"):
        dumped: dict[str, Any] = feature_collection.model_dump()
        return dumped
    if isinstance(feature_collection, dict):
        return feature_collection
    raise TypeError(f"Expected a FeatureCollection or mapping, got {type(feature_collection)}")


def bounding_box(feature_collection: Any) -> BoundingBox | None:
    """Compute the GeoJSON bbox [west, south, east, north] of a FeatureCollection.

    Accepts a geojson-pydantic FeatureCollection or a plain GeoJSON mapping. Returns
    None when there are no coordinates to bound (no features, or all geometries empty
    or missing).
    """
    data = _as_dict(feature_collection)
    longitudes: list[float] = []
    latitudes: list[float] = []
    for feature in data.get("features") or []:
        if not isinstance(feature, dict):
            continue
        for lon, lat in _geometry_positions(feature.get("geometry")):
            longitudes.append(lon)
            latitudes.append(lat)

    if not longitudes:
        return None
    return (min(longitudes), min(latitudes), max(longitudes), max(latitudes))

File: frontend/src/lib/geo.ts
// Minimal GeoJSON types, bbox derivation, and a sequential color ramp for the
// MapLibre choropleth. Geometry projection is handled by MapLibre itself.

export type BBox = [number, number, number, number] // [west, south, east, north]

export interface Geometry {
  type: string
  coordinates?: unknown
  geometries?: Geometry[]
}

export interface Feature {
  type: 'Feature'
  geometry: Geometry | null
  properties: Record<string, unknown> | null
}

export interface FeatureCollection {
  type: 'FeatureCollection'
  features: Feature[]
  bbox?: number[]
}

/** Narrow an unknown value to a FeatureCollection after a minimal shape check. */
export function asFeatureCollection(value: unknown): FeatureCollection | null {
  if (
    typeof value === 'object' &&
    value !== null &&
    (value as FeatureCollection).type === 'FeatureCollection' &&
    Array.isArray((value as FeatureCollection).features)
  ) {
    return value as FeatureCollection
  }
  return null
}

/** Yield every [lon, lat] position from an arbitrarily nested coordinate array. */
function* positions(coordinates: unknown): Generator<[number, number]> {
  if (
    Array.isArray(coordinates) &&
    coordinates.length >= 2 &&
    typeof coordinates[0] === 'number' &&
    typeof coordinates[1] === 'number'
  ) {
    yield [coordinates[0], coordinates[1]]
    return
  }
  if (Array.isArray(coordinates)) {
    for (const item of coordinates) yield* positions(item)
  }
}

/** Every position of a geometry, recursing into GeometryCollections. */
function* geometryPositions(geometry: Geometry | null): Generator<[number, number]> {
  if (!geometry) return
  if (geometry.type === 'GeometryCollection') {
    for (const sub of geometry.geometries ?? []) yield* geometryPositions(sub)
    return
  }
  yield* positions(geometry.coordinates)
}

/** Derive a bbox from features when the collection does not carry one. */
export function bboxOfFeatures(features: Feature[]): BBox | null {
  let west = Infinity
  let south = Infinity
  let east = -Infinity
  let north = -Infinity
  for (const feature of features) {
    for (const [lon, lat] of geometryPositions(feature.geometry)) {
      if (lon < west) west = lon
      if (lon > east) east = lon
      if (lat < south) south = lat
      if (lat > north) north = lat
    }
  }
  if (!Number.isFinite(west)) return null
  return [west, south, east, north]
}

// Sequential ColorBrewer-style OrRd ramp (light -> dark), interpolated in RGB.
const RAMP: [number, number, number][] = [
  [254, 240, 217],
  [253, 204, 138],
  [252, 141, 89],
  [227, 74, 51],
  [179, 0, 0],
]

/** Map a normalized value in [0, 1] to a hex color along the sequential ramp. */
export function rampColor(t: number): string {
  const clamped = Math.max(0, Math.min(1, Number.isFinite(t) ? t : 0))
  const scaled = clamped * (RAMP.length - 1)
  const index = Math.min(RAMP.length - 2, Math.floor(scaled))
  const frac = scaled - index
  const [r1, g1, b1] = RAMP[index]
  const [r2, g2, b2] = RAMP[index + 1]
  const mix = (a: number, b: number) => Math.round(a + (b - a) * frac)
  return `rgb(${mix(r1, r2)}, ${mix(g1, g2)}, ${mix(b1, b2)})`
}
